vnext/ReactUWP/Threading/UIMessageQueueThread.cpp

```diff
@@ -17,25 +17,41 @@ UIMessageQueueThread::UIMessageQueueThread(
 UIMessageQueueThread::~UIMessageQueueThread() {}
 
 void UIMessageQueueThread::runOnQueue(std::function<void()> &&func) {
-  m_uiDispatcher.RunAsync(
-      winrt::Windows::UI::Core::CoreDispatcherPriority::Normal,
-      [func = std::move(func)]() {
+  ensureQueue();
+  m_queue->push_back(func);
 
 //#define TRACK_UI_CALLS
 #ifdef TRACK_UI_CALLS
-        char buffer[1024];
-        static uint32_t cCalls = 0;
-        _snprintf_s(
-            buffer,
-            _countof(buffer),
-            _TRUNCATE,
-            "UIMessageQueueThread Calls: %u\r\n",
-            ++cCalls);
-        OutputDebugStringA(buffer);
+  char buffer[1024];
+  static uint32_t cCalls = 0;
+  _snprintf_s(
+      buffer,
+      _countof(buffer),
+      _TRUNCATE,
+      "UIMessageQueueThread Calls: %u\r\n",
+      ++cCalls);
+  OutputDebugStringA(buffer);
 #endif
+}
+
+void UIMessageQueueThread::ensureQueue() {
+  if (m_queue == nullptr) {
+    m_queue = std::make_unique<WorkItemQueue>();
+    m_queue->reserve(2048);
+  }
+}
 
-        func();
-      });
+void UIMessageQueueThread::onBatchComplete() {
+  std::shared_ptr<WorkItemQueue> queue = m_queue;
+  m_queue = nullptr;
+  if (queue) {
+    m_uiDispatcher.RunAsync(
+        winrt::Windows::UI::Core::CoreDispatcherPriority::Normal, [queue]() {
```

Comment:
[queue{std::move(queue)}] to move not copy the queue

Reply:
avoid copying a smart pointer?  done anyway.

---
In reply to: [314420144](https://github.com/microsoft/react-native-windows/pull/2937#discussion_r314420144) [](ancestors = 314420144)